### PR TITLE
WL-5141: Ability to zip selected files/folders

### PR DESCRIPTION
**Pull request by nicholaswilson100**
The functionality this is expected to provide is that a user can navigate around their Resources, select some files and folders and then click 'Zip' which will save a zip file with the same name as their site in their Resources (we're deliberately ignoring the 'exceeding quota' problem and 'downloading the zip file for them' feature for now). Importantly, the zip file should preserve the folder structure of their resources, i.e., if they selected a file inside folder A, then the zip file should have folder A in it and the file inside that.  

This code is based partly off what the code does for the 'Move to Trash' function and also partly on the 'Compress to Zip Folder' action.  I've removed unnecessary parts or parts just to do with deleting and otherwise tried to clean it up as much as possible whilst respecting the Sakai way of doing things.

### CODE REVIEW

File: content/content-tool/tool/src/java/org/sakaiproject/content/tool/ResourcesAction.java [hunk 1]
```diff
@@ -54,6 +54,7 @@
 import java.util.regex.Pattern;
 
 import org.apache.commons.lang.StringUtils;
+import org.sakaiproject.content.util.*;
```
Comment (buckett) [inline content/content-tool/tool/src/java/org/sakaiproject/content/tool/ResourcesAction.java]: Sakai prefers expanded imports.

File: content/content-tool/tool/src/java/org/sakaiproject/content/tool/ResourcesAction.java [hunk 2]
```diff
@@ -6556,6 +6595,52 @@ public void doDeleteconfirm ( RunData data)
...
+			initCopyContext(state);
+		}
+
+		// cancel move if there is one in progress
+		if(! Boolean.FALSE.toString().equals(state.getAttribute (STATE_MOVE_FLAG)))
+		{
+			initMoveContext(state);
+		}
+
+		Set zipIdSet  = new TreeSet();
```
Comment (buckett) [inline content/content-tool/tool/src/java/org/sakaiproject/content/tool/ResourcesAction.java]: Add generics?

File: content/content-tool/tool/src/java/org/sakaiproject/content/tool/ResourcesAction.java [hunk 3]
```diff
@@ -4086,6 +4090,36 @@ else if (RESOURCES_MODE_DROPBOX.equalsIgnoreCase((String) state.getAttribute(STA
...
+
+		List zipItems = (List) state.getAttribute(STATE_DELETE_SET);
+
+		context.put ("zipItems", zipItems);
+
+		//  %%STATE_MODE_RESOURCES%%
+		//not show the public option when in dropbox mode
+		if (RESOURCES_MODE_RESOURCES.equalsIgnoreCase((String) state.getAttribute(STATE_MODE_RESOURCES)))
+		{
+			context.put("dropboxMode", Boolean.FALSE);
```
Comment (buckett) [inline content/content-tool/tool/src/java/org/sakaiproject/content/tool/ResourcesAction.java]: `dropboxMode` doesn't look to to be used in the template

File: content/content-tool/tool/src/java/org/sakaiproject/content/tool/ResourcesAction.java [hunk 4]
```diff
@@ -4086,6 +4090,36 @@ else if (RESOURCES_MODE_DROPBOX.equalsIgnoreCase((String) state.getAttribute(STA
...
+		}
+		else if (RESOURCES_MODE_DROPBOX.equalsIgnoreCase((String) state.getAttribute(STATE_MODE_RESOURCES)))
+		{
+			// not show the public option or notification when in dropbox mode
+			context.put("dropboxMode", Boolean.TRUE);
+		}
+		context.put("homeCollection", (String) state.getAttribute (STATE_HOME_COLLECTION_ID));
+		context.put("siteTitle", state.getAttribute(STATE_SITE_TITLE));
+		context.put ("resourceProperties", ContentHostingService.newResourceProperties ());
```
Comment (buckett) [inline content/content-tool/tool/src/java/org/sakaiproject/content/tool/ResourcesAction.java]: `resourceProperties` doesn't look to be used in the template.

File: content/content-tool/tool/src/java/org/sakaiproject/content/tool/ResourcesAction.java [hunk 5]
```diff
@@ -4086,6 +4090,36 @@ else if (RESOURCES_MODE_DROPBOX.equalsIgnoreCase((String) state.getAttribute(STA
...
+		{
+			context.put("dropboxMode", Boolean.FALSE);
+		}
+		else if (RESOURCES_MODE_DROPBOX.equalsIgnoreCase((String) state.getAttribute(STATE_MODE_RESOURCES)))
+		{
+			// not show the public option or notification when in dropbox mode
+			context.put("dropboxMode", Boolean.TRUE);
+		}
+		context.put("homeCollection", (String) state.getAttribute (STATE_HOME_COLLECTION_ID));
+		context.put("siteTitle", state.getAttribute(STATE_SITE_TITLE));
```
Comment (buckett) [inline content/content-tool/tool/src/java/org/sakaiproject/content/tool/ResourcesAction.java]: `siteTitle` doesn't look to be used in the template.

**Comment by nicholaswilson100** [comment]
ta

File: content/content-tool/tool/src/java/org/sakaiproject/content/tool/ResourcesAction.java [hunk 6]
```diff
@@ -7046,6 +7131,78 @@ public void doFinalizeDelete( RunData data)
...
+	public void doFinalizeZip( RunData data) throws IdUnusedException {
+		logger.debug(this + ".doFinalizeZip()");
+
+		if (!"POST".equals(data.getRequest().getMethod())) {
+			return;
+		}
+
+		SessionState state = ((JetspeedRunData)data).getPortletSessionState (((JetspeedRunData)data).getJs_peid ());
+
+		String oldCollectionId = (String) state.getAttribute(STATE_COLLECTION_ID);
```
Comment (buckett) [inline content/content-tool/tool/src/java/org/sakaiproject/content/tool/ResourcesAction.java]: `oldCollectionId` looks like a copy/paste variable name.

File: content/content-tool/tool/src/webapp/vm/content/sakai_resources_list.vm [hunk 7]
```diff
@@ -884,15 +889,15 @@
 			var counter = document.getElementById(key + "-count");
 			var button = document.getElementById(key + "-button");
 			
-			if(caller && caller.checked && caller.checked == true)
```
Comment (buckett) [inline content/content-tool/tool/src/webapp/vm/content/sakai_resources_list.vm]: Looks to be pure whitespace change.

File: kernel/kernel-util/src/main/java/org/sakaiproject/content/util/ZipContentUtil.java [hunk 8]
```diff
@@ -467,14 +475,64 @@ private void storeContentCollection(String rootId, ContentCollection collection,
...
+	 */
+	private void storeContentCollection(String rootId, ContentCollection collection, ZipOutputStream out, List<String> selectedFolders, List<String> selectedFiles) throws Exception {
+		List<String> members = collection.getMembers();
+		for (String memberId: members) {
+			if (memberId.endsWith(Entity.SEPARATOR)) {
+				if (selectedFolders != null && selectedFolders.contains(memberId)) { // user selected folder
+					ContentCollection memberCollection = ContentHostingService.getCollection(memberId);
+					storeContentCollection(rootId, memberCollection, out);
+				}
+				else if (isInPath(selectedFiles, selectedFolders, memberId)) {// it's a folder above a selected folder
```
Comment (buckett) [inline kernel/kernel-util/src/main/java/org/sakaiproject/content/util/ZipContentUtil.java]: Comment should be: `it's a folder above a selected folder or file`

File: content/content-tool/tool/src/webapp/vm/content/sakai_resources_zipFinish.vm [hunk 9]
```diff
@@ -0,0 +1,59 @@
+<!-- chef_resources_zipFinish.vm, version: $Revision:  $, use with org.sakaiproject.tool.content.ResourcesAction.java -->
+<div class="portletBody specialLink">
+	#if ($alertMessage)<div class="alertMessage">$tlang.getString("label.alert") $validator.escapeHtml($alertMessage)</div><div style="display:block;clear:both" ></div>#end
+	<div class="alertMessage">$tlang.getString("zip.sure")</div><div style="display:block;clear:both" ></div>
+	<h3>
+        $tlang.getString("zip.zip")
+	</h3>
+	<form action="#toolForm("ResourcesAction")" name="deleteFileForm" method="post" >
+		<input type="hidden" name="collectionId" value="$validator.escapeHtml($collectionId)" />
+		#if ($zipItems.size()>0)
```
Comment (buckett) [inline content/content-tool/tool/src/webapp/vm/content/sakai_resources_zipFinish.vm]: Have a `#else# clause with error, or why bother doing the test as the user will just see an empty page with no cancel button?

**Review by buckett — CHANGES_REQUESTED** [review]
Really nice feature, I can see that if you're going to stream the file back to the user then it's not worth doing the quota work.

What about dropping the old "Compress to Zip" option on the menu as well (do we need 2 ways of doing it?).

When you're walking a filesystem I would have been tempted to use something more along the style of `java.nio.file.PathMatcher` that holds the sets of files/folders to export and a simpler walker that consults the matcher as it finds files/folders (this is just a comment, this doesn't need fixing).

**Comment by nicholaswilson100** [comment]
Tweaked

**Review by buckett — APPROVED** [review]

**Comment by buckett** [comment]
Cool, is there a reason for splitting this into two parts?

**Comment by nicholaswilson100** [comment]
Only to go by the usual Agile principle 'Iterative, incremental and evolutionary'.  The idea is that you get something small useful and working out fast, show it to the stakeholders and iterate over it to get a better product. That way you maximise the likelihood of giving users what they want. i.e., I don't want to prioritse  improvements to it until the stakeholders have had a chance to use it and input their views.

**Pull request merged**